### PR TITLE
Parallel hic build matrix

### PR DESCRIPTION
**Pull request by joachimwolff**
As stated in PR #86 I closed it and reopen it with another branch.

### CODE REVIEW

**Comment by fidelram** [comment]
@joachimwolff can you to a PR to the `develop` branch instead of master?

**Comment by fidelram** [comment]
Finally is green!

File: galaxy/wrapper/hicBuildMatrix.xml [hunk 1]
```diff
@@ -119,7 +119,7 @@
             </repeat>
             <param name="binSize" value="5000"/>
             <output name="outBam" file="hicBuildMatrix_result1.bam" ftype="bam"/>
-            <output name="outFileName" file="hicBuildMatrix_result1.h5" ftype="h5" compare="sim_size"/>
+            <output name="outFileName" file="hicBuildMatrix_result1.h5" ftype="h5" compare="sim_size" delta="30000"/>
```
Comment (fidelram) [inline galaxy/wrapper/hicBuildMatrix.xml]: What is the reason for this parameter (delta)? The size of the matrix should not change.

File: galaxy/wrapper/macros.xml [hunk 2]
```diff
@@ -325,7 +325,7 @@ This tool is developed by the `Bioinformatics and Deep-Sequencing Unit`_ at the
 
 .. _Bioinformatics and Deep-Sequencing Unit: http://www3.ie-freiburg.mpg.de/facilities/research-facilities/bioinformatics-and-deep-sequencing-unit/
 .. _Max Planck Institute for Immunobiology and Epigenetics: http://www3.ie-freiburg.mpg.de
-.. _help site: https://github.com/fidelram/deepTools/wiki/
+.. _help site: https://hicexplorer.readthedocs.org
```
Comment (fidelram) [inline galaxy/wrapper/macros.xml]: Please, add yourselves to the credits. Also to the list of authors in setup.py

File: hicexplorer/hicBuildMatrix.py [hunk 3]
```diff
@@ -5,7 +5,15 @@
 import time
 from os import unlink
 import os
+import shutil
 import pysam
+from six.moves import xrange
```
Comment (fidelram) [inline hicexplorer/hicBuildMatrix.py]: I am not sure how the six package works (haven't use it before), but does this means that we need to add a new dependency to setup.py?

File: hicexplorer/hicBuildMatrix.py [hunk 4]
```diff
@@ -670,38 +646,106 @@ def main(args=None):
...
+
+
+def process_data(pMateBuffer1, pMateBuffer2, pMinMappingQuality,
+                 pRemoveSelfCircles, pRestrictionSequence, pRemoveSelfLigation, pMatrixSize,
+                 pReadPosMatrix, pRfPositions, pRefId2name,
+                 pDanglingSequences, pBinsize, pResultIndex,
+                 pQueueOut, pTemplate, pOutputBamSet, pOutputName, pCounter,
+                 pSharedBinIntvalTree, pDictBinIntervalTreeIndex, pCoverage, pCoverageIndex, pOutputFileBufferDir,
+                 pRow, pCol, pData):
+
```
Comment (fidelram) [inline hicexplorer/hicBuildMatrix.py]: Can you add a docstring for this function?

File: galaxy/wrapper/hicBuildMatrix.xml (hunk 1 above)
Comment (joachimwolff) [inline galaxy/wrapper/hicBuildMatrix.xml]: To quote myself from PR #89:
'I increased the delta for the h5 matrix of the test case to 30000 --> 30kb.
The computed matrix for the SRR1956527 dataset differs ~160 kb from the one computed with the master branch, both with a size of 29 MB. Maybe this is caused by different versions of the used compression or different handling of the used internal datatypes when the data is transfered to the sparse matrix etc?'

File: hicexplorer/hicBuildMatrix.py (hunk 3 above)
Comment (joachimwolff) [inline hicexplorer/hicBuildMatrix.py]: Yes we need to add it. I have done this now.

**Comment by bgruening** [comment]
Merged in https://github.com/maxplanck-ie/HiCExplorer/pull/89

**Pull request closed without merging**